float32 rightX
float32 rightY
float32 leftX
float32 leftY
